Movie/Series page

AlekseiKolesnikov opened the pull request.

AlekseiKolesnikov committed:
Hover effect on every movie poster on the main page

AlekseiKolesnikov committed:
1. Renamed categories content options
2. Fixed Layout logic. Now I use children instead of passing props every time there. I needed it to reuse a Layout component for individual movie/tv show/person page.

AlekseiKolesnikov committed:
Made a navigation to the movie page on click

AlekseiKolesnikov committed:
Made all posters clickable and follow the new route

AlekseiKolesnikov committed:
Made a request and data set logic in the movie/tv page

AlekseiKolesnikov committed:
Store a data page in the local storage, so when you update the page you can still have data and don't request it again.

ShadowZilot commented on src/components/common/Content/MainPageContent/MainPageContent.tsx:
```diff
@@ -53,18 +47,24 @@ export const AppContent = () => {
+                    />
                 );
             }
             if (index === 2 || index === 3) {
```
Index hardcode is bad practice

ShadowZilot commented on src/components/common/Content/MovieTVCategoriesContent/MovieTVCategoriesContent.tsx:
```diff
@@ -0,0 +1,19 @@
+
+    return (
+        <div className="movie-series-page__container">
+            {windowPathName !== '/' && (
```
This is strange condition

ShadowZilot commented on src/components/common/Content/MovieTVShowPageContent/MovieTVShowPageContent.tsx:
```diff
@@ -0,0 +1,61 @@
+        if (movieTvPageData && itemId === movieTvPageData.id) {
+            const storedData = localStorage.getItem(MOVIE_PAGE_CONTENT_KEY);
+
+            if (storedData && JSON.parse(storedData).length === 0) {
+                setMovieTvPageData(JSON.parse(storedData)[0]);
+            }
```
JSOM parsing twice. JSON parsing might be long time operation so better do it once.

ShadowZilot commented on src/components/common/Content/MovieTVShowPageContent/MovieTVShowPageContent.tsx:
```diff
@@ -0,0 +1,61 @@
+import { MOVIE_PAGE_CONTENT_KEY } from "../../../../data/local-storage-keys.ts";
+import { MovieTvPageContent } from "../../../../interfaces /MoviePageResponseInterface.ts";
+
+export const MovieTVShowPageContent = () => {
+    const location = useLocation()
+    const { id, type } = location.state
+    const [movieTvPageData, setMovieTvPageData] = useState((): MovieTvPageContent => {
+        const storedData = localStorage.getItem(MOVIE_PAGE_CONTENT_KEY);
+        return storedData ? JSON.parse(storedData)[0] : null;
+    })
+
+    useEffect(() => {
+        const itemId = id ? parseInt(id) : null
+
+        if (movieTvPageData && itemId === movieTvPageData.id) {
+            const storedData = localStorage.getItem(MOVIE_PAGE_CONTENT_KEY);
+
+            if (storedData && JSON.parse(storedData).length === 0) {
+                setMovieTvPageData(JSON.parse(storedData)[0]);
+            }
+        }
+    }, [id, movieTvPageData, type])
+
+    return (
+        <div className="movie-tv-page space-between-row-center-flex">
```
Also you need to understand architecture patterns in front-end development. Here you're mapping data right in component. It's bad practice. Start you diving in architecture patterns with this article https://habr.com/ru/articles/726354/

ShadowZilot commented on src/components/common/Sliders /SliderWithDescription/SliderWithDescription.tsx:
```diff
@@ -54,6 +62,24 @@ export const SliderWithDescription = ({ sliderId, label }: { sliderId: number, l
                         <SwiperSlide
                             key={data.itemId}
                             className="small-swiper__slide start-column-top-flex"
+                            onClick={() => {
+                                requestMoviePageContent(
+                                    type,
+                                    data.itemId,
+                                    (response: MovieTvPageContent | undefined) => {
+                                        if (response) {
+                                            setMoviePageContentCallback(data.itemId, response)
+                                            navigate(useMovieTVShowsNavigation.getState().movieTVShowsRoutes.pathName, {
+                                                state: {
+                                                    id: data.itemId,
+                                                    type: type
+                                                }
+                                            })
+                                        }
+                                    }
+                                )
+                                updateMovieTVShowsNavigation(type, data.itemId.toString())
```
What if you got response false?

ShadowZilot commented on src/states/useMoviePageResponse.ts:
```diff
@@ -0,0 +1,82 @@
+    getMoviePageContent: async (type, id, pageContentResponseCallback) => {
+        const contentType =  type ? (type === 'movie' ? 'movie/' : 'tv/') : ''
+        const itemId = id !== null ? id : 0
+
+        if (itemId === 0) console.log('Movie ID was not found')
+        if (contentType === '') console.log('Movie type was not found')
+
```
Remove debug logs

ShadowZilot commented on src/states/useMoviePageResponse.ts:
```diff
@@ -0,0 +1,82 @@
+        axios.get<ApiContentPageResponse>(`${BASE_MOVIE_API_URL}${contentType}${itemId.toString()}?language=en-US`, axiosHeader)
+            .then((result) => {
+                const movieTvResponse = (): MovieTvPageContent | undefined => {
+                    const response = result.data
+                    const releaseYear = response.release_date ? response.release_date.substring(0, 4) : response.first_air_date.substring(0, 4)
```
This is unsafe code because it doesn't hande other date format

ShadowZilot commented on src/states/useMoviePageResponse.ts:
```diff
@@ -0,0 +1,82 @@
+import axios from "axios";
+import { MOVIE_PAGE_CONTENT_KEY } from "../data/local-storage-keys.ts";
+
+const API_TOKEN = "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI5MGRlZTQwNjJkZGM1NDI3NzI0ZTdjYmQ2YTg2ZTlhMSIsIm5iZiI6MTcyNTIyNjU4MS4xNTQ3NTUsInN1YiI6IjY0YjE5ZTQ2MGU0ZmM4MDBhZDVkMTA4NSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.trEXym5ftq5vCjNXU63PAobmGngKZ_xDB1nKyYqs-NE";
```
I wish you can figure out more safe way to store API key

AlekseiKolesnikov committed:
1. Changed index to sliderId in the MainPageContent.tsx
2. Removed condition from MovieTVCategoriesContent.tsx
3. Fixed styles, because they were massed up after I changed Layout logic
4. Change the logic of set data in the MovieTVShowPageContent.tsx. Also, removed double parsing from the code
5. Change the way of storing Api Key. Now it's stored in the .env file
6. Change request onClick logic. Put it in the custom hook and handle better the non-response

ShadowZilot commented on src/components/common/Content/MainPageContent/MainPageContent.tsx:
```diff
@@ -51,45 +46,57 @@ export const AppContent = () => {
-                    <SliderWithDescription sliderId={data.sliderId} label={data.label} key={`${key}-description`}/>
+                    <SliderWithDescription
+                        sliderId={data.sliderId}
+                        label={data.label}
+                        type={data.type}
+                        key={`${key}-description`}
+                    />
                 );
             }
-            if (index === 2 || index === 3) {
+            if (data.sliderId === 2 || data.sliderId === 3) {
                 return (
                     <div
-                        className="content-container__black-background-content start-column-top-flex"
+                        className="main-page-content-container__black-background-content start-column-top-flex"
                         key={key}
                     >
                         <SliderBlackBackground
                             sliderId={data.sliderId}
                             label={data.label}
+                            type={data.type}
                             key={`${key}-black-background`}
                         />
                         {index === 3 && <ProvidersBlackBackgroundGrid/>}
                     </div>
                 );
             }
-            if (index === 4 || index === 5) {
+            if (data.sliderId === 4 || data.sliderId === 5) {
                 return (
-                    <SliderGrid sliderId={data.sliderId} label={data.label} index={index} key={`${key}-description`}/>
+                    <SliderGrid
+                        sliderId={data.sliderId}
+                        label={data.label}
+                        index={index}
+                        type={data.type}
+                        key={`${key}-description`}
+                    />
```
You replaced index hardcode to ids hardcode, but this doesn't make difference. Better declare type of item as TS enum

AlekseiKolesnikov committed:
1. Change Date format to Date instead of string
2. Didn't use Enums TS. People do not consider it as a good practice, because its behavior can be unpredictable, but they do something similar by creating const object that you can see in my code.
